File: app/src/main/java/net/nonylene/photolinkviewer/Base58.java
package net.nonylene.photolinkviewer;

import java.math.BigInteger;

public class Base58 {
    private static final String alphabet = "123456789abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ";

    public static String decode(String string) {
        BigInteger result = BigInteger.valueOf(0);
        int piyo = string.length();
        BigInteger len = BigInteger.valueOf(58);
        for (int i = 0; i < piyo; i++) {
            BigInteger alpha = BigInteger.valueOf(alphabet.indexOf(string.substring(piyo - i - 1, piyo - i)));
            alpha = alpha.multiply(len.pow(i));
            result = result.add(alpha);
        }
        return String.valueOf(result);
    }
}

File: app/src/main/java/net/nonylene/photolinkviewer/tool/Base58.kt
package net.nonylene.photolinkviewer.tool

import java.math.BigInteger

object Base58 {
    private val CHARACTERS = "123456789abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ"
    private val CHARACTERS_LENGTH = BigInteger.valueOf(58)

    fun decode(encoded: String): String {
        return encoded.reversed().withIndex().fold(BigInteger.valueOf(0)) { result, iv ->
            val alpha = BigInteger.valueOf(CHARACTERS.indexOf(iv.value).toLong())
            result.add(alpha.multiply(CHARACTERS_LENGTH.pow(iv.index)))
        }.toString()
    }
}
